sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationApplicationSettingPropertySource.java

```diff
@@ -26,80 +30,108 @@
-                    .setLabelFilter(label);
+            SettingSelector settingSelector = new SettingSelector().setKeyFilter(keyFilter + "*").setLabelFilter(label);
 
             // * for wildcard match
-            List<ConfigurationSetting> settings = replicaClient.listSettings(settingSelector);
-
-            for (ConfigurationSetting setting : settings) {
-                String key = setting.getKey().trim().substring(keyFilter.length())
-                        .replace('/', '.');
-                if (setting instanceof SecretReferenceConfigurationSetting) {
-                    String entry = getKeyVaultEntry((SecretReferenceConfigurationSetting) setting);
-
-                    // Null in the case of failFast is false, will just skip entry.
-                    if (entry != null) {
-                        properties.put(key, entry);
-                    }
-                } else if (StringUtils.hasText(setting.getContentType())
-                        && JsonConfigurationParser.isJsonContentType(setting.getContentType())) {
-                    Map<String, Object> jsonSettings = JsonConfigurationParser.parseJsonSetting(setting);
-                    for (Entry<String, Object> jsonSetting : jsonSettings.entrySet()) {
-                        key = jsonSetting.getKey().trim().substring(keyFilter.length());
-                        properties.put(key, jsonSetting.getValue());
-                    }
-                } else {
-                    properties.put(key, setting.getValue());
-                }
+            processConfigurationSettings(replicaClient.listSettings(settingSelector), settingSelector.getKeyFilter(),
+                trim, keyVaultClientFactory);
+        }
+
+    }
+
+    protected void processConfigurationSettings(List<ConfigurationSetting> settings, String keyFilter,
+        List<String> trimStrings, AppConfigurationKeyVaultClientFactory keyVaultClientFactory)
+        throws JsonProcessingException {
+        for (ConfigurationSetting setting : settings) {
+            if (trimStrings == null && StringUtils.hasText(keyFilter)) {
+                trimStrings = new ArrayList<>();
+                trimStrings.add(keyFilter.substring(0, keyFilter.length() - 1));
+            }
+            String key = trimKey(setting.getKey(), trimStrings);
+
+            if (setting instanceof SecretReferenceConfigurationSetting) {
+                handleKeyVaultReference(key, (SecretReferenceConfigurationSetting) setting);
+            } else if (setting instanceof FeatureFlagConfigurationSetting
+                && FEATURE_FLAG_CONTENT_TYPE.equals(setting.getContentType())) {
+                handleFeatureFlag(key, (FeatureFlagConfigurationSetting) setting, trimStrings);
+            } else if (StringUtils.hasText(setting.getContentType())
+                && JsonConfigurationParser.isJsonContentType(setting.getContentType())) {
+                handleJson(setting, trimStrings);
```

Comment:
Why does this trimSettings need to pass down to these two ifs?

Reply:
So, it needs to be passed down in both as both by default call the json code. The json code needs it as we allow for trimming of prefixes from the fully formed JSON key name.